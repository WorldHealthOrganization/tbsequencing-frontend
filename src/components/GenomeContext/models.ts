import { IGene } from '../../services/genesApi/models';

export interface IGenomeData extends IGene {
  isSelected: boolean;
  y: number;
}
