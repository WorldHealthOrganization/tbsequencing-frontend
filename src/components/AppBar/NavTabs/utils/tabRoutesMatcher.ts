import type { Location } from 'react-router-dom';
import { RouterProps } from 'react-router-dom';
import { ITabsConfig, IDropdownConfig } from '../../tabsConfig';

export const tabRoutesMatcher = (location: Location, tabsConfig: ITabsConfig):boolean => {
  const foundRoute = tabsConfig.find((route) => location.pathname.includes(route.path));

  return Boolean(foundRoute);
};

export const getInitialTab = (location: Location, tabsConfig: ITabsConfig):number => {
  const foundIndex = tabsConfig.findIndex((tab) => location.pathname === tab.path);

  return foundIndex;
};

export const getActiveTab = (location: RouterProps['location'], tabsConfig: ITabsConfig):number => {
  if (typeof location !== 'string') {
    return tabsConfig.findIndex((tab) => location.pathname === tab.path);
  }
  return 0;
};

export const getActiveDropdown = (
  location: RouterProps['location'], 
  dropdownConfig: IDropdownConfig,
  defaultDropItem: string
): string => {
  if (typeof location !== 'string') {
    const menuElement = dropdownConfig.find((tab) => location.pathname === tab.path);
    return menuElement?.label || defaultDropItem;
  }
  return defaultDropItem;
};
