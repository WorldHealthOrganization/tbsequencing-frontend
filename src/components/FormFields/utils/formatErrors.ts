import { ErrorOption, Path } from 'react-hook-form';
import { IHttpError } from '../../../models/HttpError';

interface ITransformedError {
  message: ErrorOption['message'];
  type: ErrorOption['type'];
  name: Path<any>
}

export const formatErrors = (httpError?: IHttpError): ITransformedError[] | undefined => {
  if (!httpError?.data?.errors) return undefined;
  const { data: { errors } } = httpError;
  return errors.map((e) => ({
    type: 'custom',
    name: e.attr,
    message: e.detail,
  }));
};
