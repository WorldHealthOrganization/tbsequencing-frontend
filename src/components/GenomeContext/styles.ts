import appColors from '../../styles/colors';

export const genomeContext = {
  backgroundColor: appColors.tints.whoBlue.lightTransparent,
  marginRight: '0px',
  marginLeft: '0px',
  borderLeft: `2px solid ${appColors.secondary.whoBlue}`,
  borderRight: `2px solid ${appColors.secondary.whoBlue}`,
};

export const selectedBar = `fill: ${appColors.secondary.whoYellow}`;
export const defaultBar = `fill: ${appColors.tints.whoPurple.lighter}`;
