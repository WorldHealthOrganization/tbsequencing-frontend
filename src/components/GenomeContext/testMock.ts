import { IGene } from '../../services/genesApi/models';

const mockGene: IGene = {
  id: 1,
  locusTag: null,
  geneDbCrossrefId: 1,
  startPos: 760000,
  endPos: 761200,
  strand: 1,
  geneName: '1',
};

const mockData: IGene[] = [
  {
    id: 1,
    locusTag: null,
    geneDbCrossrefId: 1,
    startPos: 760000,
    endPos: 761200,
    strand: 1,
    geneName: '1',
  },
  {
    id: 2,
    locusTag: null,
    geneDbCrossrefId: 2,
    startPos: 759899,
    endPos: 760200,
    strand: 1,
    geneName: '2',
  },
  {
    id: 3,
    locusTag: null,
    geneDbCrossrefId: 3,
    startPos: 763000,
    endPos: 763300,
    strand: -1,
    geneName: '3',
  },
  {
    id: 4,
    locusTag: null,
    geneDbCrossrefId: 4,
    startPos: 760502,
    endPos: 760999,
    strand: 1,
    geneName: '4',
  },
  {
    id: 5,
    locusTag: null,
    geneDbCrossrefId: 5,
    startPos: 761000,
    endPos: 762000,
    strand: -1,
    geneName: '5',
  }, {
    id: 6,
    locusTag: null,
    geneDbCrossrefId: 6,
    startPos: 760460,
    endPos: 761000,
    strand: -1,
    geneName: '6',
  },
  {
    id: 7,
    locusTag: null,
    geneDbCrossrefId: 7,
    startPos: 759999,
    endPos: 761050,
    strand: 1,
    geneName: '7',
  },
  {
    id: 8,
    locusTag: null,
    geneDbCrossrefId: 8,
    startPos: 763000,
    endPos: 763325,
    strand: 1,
    geneName: '8',
  },
  {
    id: 9,
    locusTag: null,
    geneDbCrossrefId: 9,
    startPos: 761066,
    endPos: 762066,
    strand: -1,
    geneName: '9',
  },
  {
    id: 10,
    locusTag: null,
    geneDbCrossrefId: 10,
    startPos: 763000,
    endPos: 763112,
    strand: 1,
    geneName: '10',
  },
];

const dataSet1 = [
  {
    startPos: 1,
    endPos: 10,
  },
  {
    startPos: 2,
    endPos: 15,
  },
  {
    startPos: 20,
    endPos: 40,
  },
];

const result1 = [
  {
    y: 1,
    startPos: 1,
    endPos: 10,
  },
  {
    y: 2,
    startPos: 2,
    endPos: 15,
  },
  {
    y: 1,
    startPos: 20,
    endPos: 40,
  },
];

const dataSet2 = [
  {
    startPos: 1,
    endPos: 10,
  },
  {
    startPos: 2,
    endPos: 9,
  },
  {
    startPos: 3,
    endPos: 40,
  },
];

const result2 = [
  {
    y: 1,
    startPos: 1,
    endPos: 10,
  },
  {
    y: 2,
    startPos: 2,
    endPos: 9,
  },
  {
    y: 3,
    startPos: 3,
    endPos: 40,
  },
];

const dataSet3 = [
  {
    startPos: 1,
    endPos: 10,
  },
  {
    startPos: 20,
    endPos: 22,
  },
  {
    startPos: 40,
    endPos: 50,
  },
];

const result3 = [
  {
    y: 1,
    startPos: 1,
    endPos: 10,
  },
  {
    y: 1,
    startPos: 20,
    endPos: 22,
  },
  {
    y: 1,
    startPos: 40,
    endPos: 50,
  },
];

const dataSet4 = [
  {
    startPos: 1,
    endPos: 10,
  },
  {
    startPos: 10,
    endPos: 20,
  },
  {
    startPos: 20,
    endPos: 50,
  },
];

const result4 = [
  {
    y: 1,
    startPos: 1,
    endPos: 10,
  },
  {
    y: 2,
    startPos: 10,
    endPos: 20,
  },
  {
    y: 1,
    startPos: 20,
    endPos: 50,
  },
];

const dataSet5 = [
  {
    startPos: 50,
    endPos: 80,
  },
  {
    startPos: 30,
    endPos: 50,
  },
  {
    startPos: 200,
    endPos: 250,
  },
];

const result5 = [
  {
    y: 1,
    startPos: 30,
    endPos: 50,
  },
  {
    y: 2,
    startPos: 50,
    endPos: 80,
  },
  {
    y: 1,
    startPos: 200,
    endPos: 250,
  },
];

export {
  dataSet1,
  dataSet2,
  dataSet3,
  dataSet4,
  dataSet5,
  result1,
  result2,
  result3,
  result4,
  result5,
  mockData,
  mockGene,
};
