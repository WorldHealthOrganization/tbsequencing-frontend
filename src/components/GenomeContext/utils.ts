import { IGene } from '../../services/genesApi/models';

const DEFAULT_Y_AXES = 1;

export const formatGenomeContextData = (data: any[], selectedItem?: IGene) => {
  const dataWithYAxes = data.map((item) => ({
    ...item,
    y: DEFAULT_Y_AXES,
  }));

  const sortedData = dataWithYAxes.sort((a, b) => {
    if (a.startPos < b.startPos) return -1;
    if (a.startPos > b.startPos) return 1;
    return 0;
  });

  for (let i = 0; i < sortedData.length; i += 1) {
    const currItem = sortedData[i];
    const nextItem = sortedData[i + 1];
    const prevItem = sortedData[i - 1];

    if (selectedItem && currItem.geneDbCrossrefId === selectedItem.geneDbCrossrefId) {
      currItem.isSelected = true;
    }

    if (!nextItem) {
      break;
    }

    const nextIncludesCurr = nextItem.startPos <= currItem.endPos;
    const prevNotIncludeNext = prevItem && (nextItem.startPos > prevItem.endPos);

    if (nextIncludesCurr) {
      nextItem.y = currItem.y + 1;
    }

    if (prevNotIncludeNext) {
      if (currItem.y !== prevItem.y) nextItem.y = prevItem.y;
    }
  }

  return sortedData;
};
