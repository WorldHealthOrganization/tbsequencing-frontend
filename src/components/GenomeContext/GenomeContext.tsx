/** @jsxImportSource @emotion/react */
import {
  scaleBand, scaleLinear,
} from 'd3-scale';

import React from 'react';
import { max } from 'd3-array';
import { useD3 } from '../../hooks/useD3';
import { defaultBar, genomeContext, selectedBar } from './styles';
import { font } from '../typography/fontSizes';
import { formatGenomeContextData } from './utils';
import { IGenomeData } from './models';
import { IGenomeContextProps } from '../../features/genesView/models';

// EXAMPLE ARROW TO THE LEFT
// 20,20 5,35.5 20,52 100,52 100,20
// {
//   1: left-bottom coordinate (x1 + 15),
//   2: arrow-dot coordinate (x1),
//   3: left-top coordinate (x1 + 15),
//   4: right-top coordinate (x2),
//   5: right-bottom coordinate (x2),
// }

// EXAMPLE ARROW TO THE RIGHT
// 100,52 20,52 20,20 100,20 115,35
// {
//   100,52 = right-top coordinate (x2 - 15),
//   20,52 = left-top coordinate (x1),
//   20,20 = left-bottom coordinate (x1),
//   100,20 = right-bottom coordinate (x2 - 15),
//   115,35 = arrow-dot coordinate (x2)
// }

// polygon points it is an array with pairs of coordinates x,y
// x1 and x2 calculates by d3 library, depends on passed range,
// so we just pass x's in and change X coordinates from default example
// Y coordinates remains same
const getPolygonPoints = (x1: number, x2: number, strand?: 1 | -1) => {
  // arrow to the left
  if (strand === -1) {
    // if gene is to small need to add arrow not to transform it
    if (x1 + 15 > x2) {
      return `${x1},20 ${x1 - 15},35.5 ${x1},52 ${x2},52 ${x2},20`;
    }
    return `${x1 + 15},20 ${x1},35.5 ${x1 + 15},52 ${x2},52 ${x2},20`;
  }
  // arrow to the right

  // if gene is to small need to add arrow not to transform it
  if (x2 - 15 < x1) {
    return `${x2},52 ${x1},52 ${x1},20 ${x2},20 ${x2 + 15},35.5`;
  }
  return `${x2 - 15},52 ${x1},52 ${x1},20 ${x2 - 15},20 ${x2},35.5`;
};

export const GenomeContext = ({
  data, selectedGene, minDomain, maxDomain, width,
}: IGenomeContextProps) => {
  const ref = useD3((svg) => {
    if (!data?.length) return;

    const formattedData = formatGenomeContextData(data, selectedGene);

    const height = max(formattedData, (d) => d.y) * 65;
    const keyFunction = (d: IGenomeData) => d.geneDbCrossrefId;

    const x = scaleLinear([minDomain!, maxDomain!], [0, width]);
    const y = scaleBand()
      .domain(formattedData.map((item) => item.y as unknown as string))
      .rangeRound([0, height]);

    const rectTransform = (d: IGenomeData) => `translate(0, ${y(d.y as unknown as string)})`;

    // Clear everything before rerender
    svg.selectAll('*').remove();

    svg
      .attr('class', 'chart')
      .attr('width', width)
      .attr('height', height);

    const group = svg.selectAll('.chart')
      .data(formattedData, keyFunction).enter()
      .append('g')
      .attr('y', 0)
      .attr('style', (d: IGenomeData) => {
        if (d.isSelected) return selectedBar;
        return defaultBar;
      })
      .attr('transform', rectTransform)
      .attr('height', 32);

    group.append('polygon')
      .attr('style', (d: IGenomeData) => {
        if (d.isSelected) return selectedBar;
        return defaultBar;
      })
      .attr('points', (d: IGenomeData) => getPolygonPoints(x(d.startPos), x(d.endPos), d.strand));

    group.append('text')
      .text((d: IGenomeData) => d.geneName || d.locusTag)
      .attr('font-size', 16)
      .attr('font-family', font.bold)
      .attr('fill', 'black')
      .attr('y', 36)
      .attr('x', (d: IGenomeData) => {
        // If gene startPos less than min coordinate,
        // or gene endPos more than end coordinate,
        // it means that gene overflows the widget, and
        // we have to cut the polygon to display
        // locus tag || geneName relative to min or max coordinate accordingly
        const start = d.startPos < minDomain! ? minDomain! : d.startPos;
        const end = d.endPos > maxDomain! ? maxDomain! : d.endPos;
        return x((start + end) / 2);
      })
      .attr('dominant-baseline', 'middle')
      .attr('text-anchor', 'middle');

    svg.exit().remove();
  }, [data, minDomain, maxDomain, selectedGene, width]);

  return (
    <svg
      style={genomeContext}
      ref={ref}
    />
  );
};

export default GenomeContext;
