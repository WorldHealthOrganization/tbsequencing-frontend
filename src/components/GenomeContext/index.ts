import GenomeContext from './GenomeContext';

export default GenomeContext;
