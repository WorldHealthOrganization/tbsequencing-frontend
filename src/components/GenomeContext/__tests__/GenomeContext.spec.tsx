import React from 'react';
import { getTreeJSONWithRedux, render } from '../../../utils/testUtils/jestUtils';
import GenomeContext from '../GenomeContext';
import {
  dataSet1,
  dataSet2,
  dataSet3,
  dataSet4,
  dataSet5,
  result1,
  result2,
  result3,
  result4,
  result5,
  mockData, mockGene,
} from '../testMock';
import { formatGenomeContextData } from '../utils';

describe('GenomeContext component test', () => {
  test('renders GenomeContext component without error', () => {
    render(
      <GenomeContext
        isLoading={false}
        data={mockData}
        selectedGene={mockGene}
        maxDomain={760000}
        minDomain={763112}
      />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <GenomeContext
        isLoading={false}
        data={mockData}
        selectedGene={mockGene}
        maxDomain={760000}
        minDomain={763112}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('GenomeContext layout', () => {
  test.each([
    [dataSet1, result1],
    [dataSet2, result2],
    [dataSet3, result3],
    [dataSet4, result4],
    [dataSet5, result5],
  ])('successfully format data to render rows', (data, expected) => {
    const result = formatGenomeContextData(data);

    expect(result).toEqual(expected);
  });
});
