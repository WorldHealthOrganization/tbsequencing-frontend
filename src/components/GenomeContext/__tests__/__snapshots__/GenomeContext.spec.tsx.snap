// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenomeContext component test Should render correctly 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "#E8F1F6",
      "borderLeft": "2px solid #009ADE",
      "borderRight": "2px solid #009ADE",
      "marginLeft": "0px",
      "marginRight": "0px",
    }
  }
/>
`;
